pkg/providers/vsphere/virtualmachine/cleanup.go

```diff
@@ -0,0 +1,215 @@
+	removeVMOperatorExtraConfig(moVM.Config, &configSpec)
+
+	// Clear ManagedBy field
+	clearManagedBy(moVM.Config, &configSpec)
+
+	// Remove tag associations
+	if pkgcfg.FromContext(vmCtx).Features.VSpherePolicies {
+		removeTagAssociations(moVM.Config, &configSpec)
+	}
+
+	// Apply the configuration changes
+	if err := reconfigureVM(vmCtx, vcVM, configSpec); err != nil {
+		return fmt.Errorf("failed to apply cleanup configuration: %w", err)
+	}
+
+	vmCtx.Logger.Info("Successfully sanitized the vCenter VM")
+	return nil
+}
+
+// removeTagAssociations removes any tags that were attached to the VM
+// by VM operator.
+func removeTagAssociations(
+	config *vimtypes.VirtualMachineConfigInfo,
+	configSpec *vimtypes.VirtualMachineConfigSpec) {
+
+	if config == nil {
+		return
+	}
+
+	// VM operator associates tags to VMs that specify either
+	// mandatory or optional compute policy. Remove those from the
+	// VMs.
+	// VM operator stores any tag that it applies to the VM in the
+	// ExtraConfig.
+	ec := object.OptionValueList(config.ExtraConfig)
+	if v, _ := ec.GetString(policy.ExtraConfigPolicyTagsKey); v != "" {
+		tagSpecs := []vimtypes.TagSpec{}
+		for tag := range strings.SplitSeq(v, ",") {
+			// Just because a tag is present in the ExtraConfig
+			// does not guarantee that it is also associated to
+			// the VM. However, talking to tagging service for
+			// each tag might be non-performant. And Reconfigure
+			// will ignore these tags anyway.
+			ts := vimtypes.TagSpec{
+				ArrayUpdateSpec: vimtypes.ArrayUpdateSpec{
+					Operation: vimtypes.ArrayUpdateOperationRemove,
+				},
+				Id: vimtypes.TagId{
+					Uuid: tag,
+				},
+			}
+			tagSpecs = append(tagSpecs, ts)
+		}
+
+		configSpec.TagSpecs = append(configSpec.TagSpecs, tagSpecs...)
+	}
+}
+
+// removeVMOperatorExtraConfig identifies and marks for removal all
+// ExtraConfig keys that were added by VM Operator.
+//
+// Unlike FilteredExtraConfig which preserves allow-listed keys for reuse
+// scenarios, this method removes ALL VM Operator keys (both deny-listed
+// and allow-listed) since we're unregistering the VM from Supervisor.
+func removeVMOperatorExtraConfig(
+	config *vimtypes.VirtualMachineConfigInfo,
+	configSpec *vimtypes.VirtualMachineConfigSpec) {
+
+	if config == nil {
+		return
+	}
+
+	currentEC := object.OptionValueList(config.ExtraConfig)
+
+	// Get the deny prefixes and allow keys from extraconfig.go
+	// We need to remove keys matching BOTH deny prefixes AND allow list
+	denyPrefixes := []string{
+		"guestinfo.",
+		"vmservice.",
+		"imageregistry.",
+	}
+
+	for _, optVal := range currentEC {
+		if opt := optVal.GetOptionValue(); opt != nil {
+			key := opt.Key
+
+			// Clean up the key if it has a denyPrefix. But crucially,
+			// also clean up if the key is allowed since this is a
+			// total unregister of a VM.
+			denyPrefixMatch := slices.ContainsFunc(denyPrefixes, func(prefix string) bool {
+				return strings.HasPrefix(key, prefix)
```

Comment:
Consider checking this with case-insensitive to make it consistent with the filter in extraconfig.go: https://github.com/vmware-tanzu/vm-operator/blob/a5e6ef2ddd04269a49724c339a3b1b0a5e65c026/pkg/providers/vsphere/virtualmachine/extraconfig.go#L63